ngsPETSc/utils/firedrake/adaptive_hierarchy.py

```diff
@@ -0,0 +1,307 @@
+        ]
+        c2f_adjusted = {
+            j: np.zeros((n, j))
+            for n, j in zip(n, list(range(1, max_children + 1)))
+            if n != 0
+        }
+        f2c_adjusted = {
+            j: np.zeros((n * j, 1))
+            for n, j in zip(n, list(range(1, max_children + 1)))
+            if n != 0
+        }
+
+        coarse_full_to_sub_map = {
+            i: full_to_sub(coarse_mesh, c_subm[i], i) for i in c_subm
+        }
+        fine_full_to_sub_map = {
+            j: full_to_sub(mesh, f_subm[j], int("10" + str(j))) for j in f_subm
+        }
+
+        for i in range(len(c2f)):
+            n = len(c2f[i])
+            if 1 <= n <= max_children:
+                c2f_adjusted[n][coarse_full_to_sub_map[n](i)] = fine_full_to_sub_map[n](
+                    np.array(c2f[i])
+                )
+
+        for j in range(len(f2c)):
+            n = int(num_children[f2c[j]].item())
+            if 1 <= n <= max_children:
+                f2c_adjusted[n][fine_full_to_sub_map[n](j), 0] = coarse_full_to_sub_map[
+                    n
+                ](f2c[j].item())
+
+        c2f_subm = {
+            i: {Fraction(0, 1): c2f_adjusted[i].astype(int)} for i in c2f_adjusted
+        }
+        f2c_subm = {i: {Fraction(1, 1): f2c_adjusted[i]} for i in f2c_adjusted}
+
+        hierarchy_dict = {
+            i: HierarchyBase(
+                [c_subm[i], f_subm[i]], c2f_subm[i], f2c_subm[i], nested=True
+            )
+            for i in c_subm
+        }
+        self.submesh_hierarchies.append(hierarchy_dict)
+
+    def refine(self, refinements):
+        # Input a list corresponding to which elements get refined
+        ngmesh = self.meshes[-1].netgen_mesh
+        for l, el in enumerate(ngmesh.Elements2D()):
+            el.refine = 0
+            if refinements[l] == 1:
+                el.refine = 1
+
+        ngmesh.Refine(adaptive=True)
+        mesh = Mesh(ngmesh)
+        self.add_mesh(mesh)
+
+    def adapt(self, eta, theta):
+        # Implement Dorfler marking, returns new mesh
+        mesh = self.meshes[-1]
+        W = FunctionSpace(mesh, "DG", 0)
+        markers = Function(W)
+
+        with eta.dat.vec_ro as eta_:
+            eta_max = eta_.max()[1]
+
+        should_refine = conditional(gt(eta, theta * eta_max), 1, 0)
+        markers.interpolate(should_refine)
+
+        refined_mesh = mesh.refine_marked_elements(markers)
+        self.add_mesh(refined_mesh)
+        return refined_mesh
+
+    def split_function(self, u, child=True):
+        # Split function across submeshes
+        V = u.function_space()
+        full_mesh = V.mesh()
+        _, level = get_level(full_mesh)
+
+        ind = 1 if child else 0
+        hierarchy_dict = self.submesh_hierarchies[int(level) - ind]
+        u_corr_space = Function(
+            V.reconstruct(
+                hierarchy_dict[[*hierarchy_dict][0]].meshes[ind].submesh_parent
+            ),
+            val=u.dat,
+        )
```

Comment:
Could this be broken into 3 lines (get the submesh_parent, get the space, and the Function)